File: linkedList.py
class Node:
    def __init__(self,data, next=None):
        self.data = data
        self.next = next # 두번째 인자가 없다면 DEFAULT 인 none 이 들어감

def add(data):
    node = head
    while node.next:
        node = node.next
    # 마지막 노드가 node 가 node 가 된다
    node.next = Node(data) # 마지막 노드의 포인터에 Node 저장

def printNode(head):
    node = head
    while node.next:
        print(node.data)
        node = node.next
    print(node.data)

def searchNode(head, target):
    node = head
    search = True
    while search:
        if node.data <= target.data:
            search = False # 만약 타겟이 해당 데이터 보다 작다면 탈출
        else:
            node = node.next
    node.next, target.next = target, node.next



node1 = Node(1)
node3 = Node(1.5)
# node2 = Node(2)
# node1.next = node2
head = node1
for i in range(2,10):
    add(i)
# printNode(head)

searchNode(node1, node3)
printNode(head)

File: LinkedList.java


public class LinkedList {
    public static class Node {

        Node next;
        String data;

        public Node(String data) {
            this.data = data;
        }

    }

    Node head;

    public void addFirst(String data) {
        if(head == null) {
            head = new Node(data);
            return;
        }
        Node newHead = new Node(data);
        newHead.next = head;
        head = newHead;
    }

    public void addLast(String data) {
        if(head == null) {
            head = new Node(data);
            return;
        }
        Node current = head;
        while(current.next != null) {
            current = current.next;
        }
        current.next = new Node(data);
    }

    public void removeFirst() {
        if(head == null) {
            return;
        }
        head = head.next;
    }

    public void removeLast() {
        if(head == null) {
            return;
        }
        if(head.next == null) {
            head = null;
            return;
        }
        Node current = head;
        Node previous = null;
        while(current.next != null) {
            previous = current;
        }
        current = current.next;
        previous.next = null;
    }

    public void removeValue(String data) {
        if(head == null) {
            return;
        }
        if(head.data.equals(data)) {
            head = head.next;
            return;
        }
        Node current = head;
        while(current.next != null) {
            if(current.next.data.equals(data)) {
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }
    }
}

